pkg/lock/init_waitgroup.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package lock
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cilium/hive/cell"
+)
+
+// NewInitWaitGroup constructs a new [InitWaitGroup].
+func NewInitWaitGroup(lc cell.Lifecycle) *InitWaitGroup {
+	iwg := &InitWaitGroup{
+		initializers: map[string]InitWaitFunc{},
+	}
+	lc.Append(iwg)
+	return iwg
+}
+
+// InitWaitFunc is a function for waiting until some initialization has completed.
+// If the context given to it is cancelled the function should stop and return ctx.Err().
+type InitWaitFunc = func(context.Context) error
+
+// InitWaitGroup is a wait group construct waiting for initialization of a dynamic
+// set of components registered during Hive population time.
+//
+// It is conceptually similar to [StoppableWaitGroup], but differs in that it
+// enforces that Add() is only called before Hive has started and that initializers
+// can return errors.
+type InitWaitGroup struct {
+	mu           Mutex
+	started      bool
+	initializers map[string]InitWaitFunc
+}
+
+// Add a named initializer to block the wait group.
+// The provided function should block until initialized or until context has been
+// cancelled. It should return [context.Err] if context is cancelled.
+//
+// This method will panic if called after Hive has started to ensure that when
+// Wait() is called all initializers have already been registered.
+func (w *InitWaitGroup) Add(name string, waitFn InitWaitFunc) {
```

Comment:
What about having `Add(name)` return a `doneFunc` that is to be called when setup is complete? Or even just `Add(name)`, `Done(name)`? Might be a bit simpler.

Reply:
This was what I initially did, but then realized most places fit the ”func(Context) error” pattern better. If we’d return a done func we’d need to fork goroutines to wait and call it. We also get nicer stack trace this way.